src/AddressDerivation.hs

```diff
@@ -0,0 +1,215 @@
+    minBound = Index minBound
+    maxBound = Index maxBound
+
+instance Enum (Index 'Hardened level) where
+    fromEnum (Index ix) = fromIntegral ix
+    toEnum ix
+        | Index (fromIntegral ix) < minBound @(Index 'Hardened _) =
+            error "Index@Hardened.toEnum: bad argument"
+        | otherwise =
+            Index (fromIntegral ix)
+
+instance Enum (Index 'Soft level) where
+    fromEnum (Index ix) = fromIntegral ix
+    toEnum ix
+        | Index (fromIntegral ix) > maxBound @(Index 'Soft _) =
+            error "Index@Soft.toEnum: bad argument"
+        | otherwise =
+            Index (fromIntegral ix)
+
+instance Enum (Index 'WholeDomain level) where
+    fromEnum (Index ix) = fromIntegral ix
+    toEnum ix
+        | Index (fromIntegral ix) > maxBound @(Index 'WholeDomain _) =
+            error "Index@WholeDomain.toEnum: bad argument"
+        | otherwise =
+            Index (fromIntegral ix)
+
+instance Buildable (Index derivationType level) where
+    build (Index ix) = fromString (show ix)
+
+
+-- | Type of derivation that should be used with the given indexes.
+--
+-- In theory, we should only consider two derivation types: soft and hard.
+--
+-- However, historically, addresses in Cardano used to be generated across the
+-- both soft and hard domain. We therefore introduce a 'WholeDomain' derivation
+-- type that is the exact union of `Hardened` and `Soft`.
+data DerivationType = Hardened | Soft | WholeDomain
+
+-- | An interface for doing hard derivations from the root private key
+class HardDerivation (key :: Depth -> * -> *) where
+    type AddressIndexDerivationType key :: DerivationType
+
+    -- | Derives account private key from the given root private key, using
+    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+    -- package for more details).
+    --
+    -- NOTE: The caller is expected to provide the corresponding passphrase (and
+    -- to have checked that the passphrase is valid). Providing a wrong passphrase
+    -- will not make the function fail but will instead, yield an incorrect new
+    -- key that doesn't belong to the wallet.
+    deriveAccountPrivateKey
+        :: ScrubbedBytes
+        -> key 'RootK XPrv
+        -> Index 'Hardened 'AccountK
+        -> key 'AccountK XPrv
+
+    -- | Derives address private key from the given account private key, using
+    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+    -- package for more details).
+    --
+    -- It is preferred to use 'deriveAddressPublicKey' whenever possible to avoid
+    -- having to manipulate passphrases and private keys.
+    --
+    -- NOTE: The caller is expected to provide the corresponding passphrase (and
+    -- to have checked that the passphrase is valid). Providing a wrong passphrase
+    -- will not make the function fail but will instead, yield an incorrect new
+    -- key that doesn't belong to the wallet.
```

Comment:
We might want to change this in this library ^^"...

Reply:
removed comments